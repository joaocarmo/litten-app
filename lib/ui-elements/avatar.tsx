import { memo, useCallback } from 'react'

import { StyleSheet, View } from 'react-native'
import UIImage from 'ui-elements/image'
import {
  USER_AVATAR_SIZE_LARGE,
  USER_AVATAR_SIZE_MEDIUM,
  USER_AVATAR_SIZE_MINI,
} from 'utils/constants'
type UIAvatarProps = {
  containerStyle: any
  resizeMode: string
  size: 'mini' | 'medium' | 'large'
  source:
    | string
    | {
        uri: string
      }
  style: any
}

const areEqual = (prevProps, nextProps) => {
  const prevSource =
    typeof prevProps.source === 'string'
      ? prevProps.source
      : prevProps.source.uri
  const nextSource =
    typeof nextProps.source === 'string'
      ? nextProps.source
      : nextProps.source.uri
  return (
    prevProps.resizeMode === nextProps.resizeMode &&
    prevProps.size === nextProps.size &&
    prevSource === nextSource
  )
}

const UIAvatar: (props: UIAvatarProps) => Node = ({
  containerStyle,
  resizeMode = 'cover',
  size = 'mini',
  source,
  style,
  ...otherProps
}) => {
  const getSize = useCallback(() => {
    switch (size) {
      case 'large':
        return styles.uiAvatarSizeLarge

      case 'medium':
        return styles.uiAvatarSizeMedium

      default:
        return styles.uiAvatarSizeMini
    }
  }, [size])
  return (
    <View style={[containerStyle, styles.uiAvatarImageContainer, getSize()]}>
      {source && (
        <UIImage
          {...{ ...otherProps, resizeMode, source, style: [getSize(), style] }}
        />
      )}
    </View>
  )
}

const styles = StyleSheet.create({
  uiAvatarImageContainer: {
    overflow: 'hidden',
    alignItems: 'center',
    justifyContent: 'center',
  },
  uiAvatarSizeLarge: {
    height: USER_AVATAR_SIZE_LARGE,
    width: USER_AVATAR_SIZE_LARGE,
    borderRadius: USER_AVATAR_SIZE_LARGE,
  },
  uiAvatarSizeMedium: {
    height: USER_AVATAR_SIZE_MEDIUM,
    width: USER_AVATAR_SIZE_MEDIUM,
    borderRadius: USER_AVATAR_SIZE_MEDIUM,
  },
  uiAvatarSizeMini: {
    height: USER_AVATAR_SIZE_MINI,
    width: USER_AVATAR_SIZE_MINI,
    borderRadius: USER_AVATAR_SIZE_MINI,
  },
})
export default memo<UIAvatarProps>(
  UIAvatar,
  areEqual,
) as React$AbstractComponent<UIAvatarProps, unknown>
