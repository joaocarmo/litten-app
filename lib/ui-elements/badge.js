/**
 * @format
 * @flow
 */

import type { Node } from 'react'
import { Text, View } from 'react-native'
import { useTheme } from 'hooks'
import {
  UI_BADGE_ELEMENT_OFFSET,
  UI_BADGE_PADDING,
  UI_BADGE_SIZE,
} from 'utils/constants'

const UIBadge: (args: any) => Node = ({
  children,
  number,
  showNonPositive = false,
}) => {
  const { createStyles, typography } = useTheme()

  const styles = createStyles((theme) => ({
    uiBadgeContainer: {
      position: 'relative',
    },
    uiBadgeNumber: {
      position: 'absolute',
      alignItems: 'center',
      justifyContent: 'center',
      minHeight: UI_BADGE_SIZE,
      minWidth: UI_BADGE_SIZE,
      top: -UI_BADGE_ELEMENT_OFFSET / 2,
      right: -UI_BADGE_ELEMENT_OFFSET,
      padding: UI_BADGE_PADDING,
      borderRadius: UI_BADGE_SIZE,
      backgroundColor: theme.colors.danger,
      zIndex: 1,
    },
    uiBadgeNumberText: {
      color: theme.colors.textAlt,
      fontSize: typography.fontSize.xxsmall,
      fontWeight: typography.fontWeight.bolder,
      textAlign: 'center',
    },
  }))

  return (
    <View style={styles.uiBadgeContainer}>
      {Number.isInteger(number) &&
        (showNonPositive || (!showNonPositive && number > 0)) && (
          <View style={styles.uiBadgeNumber}>
            <Text style={styles.uiBadgeNumberText}>{number}</Text>
          </View>
        )}
      {children}
    </View>
  )
}

export default UIBadge
