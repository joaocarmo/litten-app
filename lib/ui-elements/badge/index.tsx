import { Text, View } from 'react-native'
import { useTheme } from '@hooks'
import uiBadgeStyles from '@ui-elements/badge/index.styles'

const UIBadge = ({ children, number, showNonPositive = false }) => {
  const { createStyles } = useTheme()

  const styles = createStyles(uiBadgeStyles)

  return (
    <View style={styles.uiBadgeContainer}>
      {Number.isInteger(number) &&
        (showNonPositive || (!showNonPositive && number > 0)) && (
          <View style={styles.uiBadgeNumber}>
            <Text style={styles.uiBadgeNumberText}>{number}</Text>
          </View>
        )}
      {children}
    </View>
  )
}

export default UIBadge
