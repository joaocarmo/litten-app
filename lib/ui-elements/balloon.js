/**
 * @format
 * @flow
 */

import type { Node } from 'react'
import { Pressable, StyleSheet, View } from 'react-native'
import UIText from 'ui-elements/text'
import colors from 'styles/colors'
import { fontWeight } from 'styles/typography'

const UIBalloon: (args: any) => Node = ({
  children,
  style,
  type,
  ...otherProps
}) => {
  const { onPress, onLongPress } = otherProps

  const getColorScheme = () => {
    switch (type) {
      case 'error':
        return {
          container: styles.uiBalloonError,
          text: styles.uiBalloonNormalError,
        }
      case 'info':
        return {
          container: styles.uiBalloonInfo,
          text: styles.uiBalloonNormalInfo,
        }
      default:
        return {
          container: styles.uiBalloonNormal,
          text: styles.uiBalloonNormalText,
        }
    }
  }

  const baloonStyle = getColorScheme()

  const Container =
    typeof onPress === 'function' || typeof onLongPress === 'function'
      ? Pressable
      : View

  return (
    <Container style={[styles.uiBalloon, baloonStyle.container, style]}>
      {typeof children === 'string' ? (
        <UIText
          style={[styles.uiBalloonText, baloonStyle.text]}
          {...otherProps}>
          {children}
        </UIText>
      ) : (
        children
      )}
    </Container>
  )
}

const styles = StyleSheet.create({
  uiBalloon: {
    padding: 12,
    borderRadius: 12,
    borderWidth: StyleSheet.hairlineWidth,
  },
  uiBalloonError: {
    borderColor: colors.darkRed,
    backgroundColor: colors.lightRed,
  },
  uiBalloonNormalError: {
    color: colors.darkRed,
  },
  uiBalloonInfo: {
    borderColor: colors.lightBlue,
    backgroundColor: colors.lighterBlue,
  },
  uiBalloonNormalInfo: {
    color: colors.black,
  },
  uiBalloonNormal: {
    borderColor: colors.darkGray,
    backgroundColor: colors.lightGray,
  },
  uiBalloonNormalText: {
    color: colors.black,
  },
  uiBalloonText: {
    fontWeight: fontWeight.light,
  },
})

export default UIBalloon
