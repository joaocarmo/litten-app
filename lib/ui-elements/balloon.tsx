import { Pressable, StyleSheet, View } from 'react-native'
import { useTheme } from 'hooks'
import UIText from 'ui-elements/text'

const UIBalloon = ({ children, style, type, ...otherProps }) => {
  const { createStyles } = useTheme()
  const { onPress, onLongPress } = otherProps
  const styles = createStyles((theme, typography) => ({
    uiBalloon: {
      padding: 12,
      borderRadius: 12,
      borderWidth: StyleSheet.hairlineWidth,
    },
    uiBalloonError: {
      borderColor: theme.colors.dangerDark,
      backgroundColor: theme.colors.dangerLight,
    },
    uiBalloonNormalError: {
      color: theme.colors.dangerDark,
    },
    uiBalloonInfo: {
      borderColor: theme.colors.secondaryLight,
      backgroundColor: theme.colors.secondaryLighter,
    },
    uiBalloonNormalInfo: {
      color: theme.colors.neutralDark,
    },
    uiBalloonNormal: {
      borderColor: theme.colors.neutralDark,
      backgroundColor: theme.colors.neutralLight,
    },
    uiBalloonNormalText: {
      color: theme.colors.text,
    },
    uiBalloonText: {
      fontWeight: typography.fontWeight.light,
    },
  }))

  const getColorScheme = () => {
    switch (type) {
      case 'error':
        return {
          container: styles.uiBalloonError,
          text: styles.uiBalloonNormalError,
        }

      case 'info':
        return {
          container: styles.uiBalloonInfo,
          text: styles.uiBalloonNormalInfo,
        }

      default:
        return {
          container: styles.uiBalloonNormal,
          text: styles.uiBalloonNormalText,
        }
    }
  }

  const baloonStyle = getColorScheme()
  const Container =
    typeof onPress === 'function' || typeof onLongPress === 'function'
      ? Pressable
      : View
  return (
    <Container style={[styles.uiBalloon, baloonStyle.container, style]}>
      {typeof children === 'string' ? (
        <UIText
          style={[styles.uiBalloonText, baloonStyle.text]}
          {...otherProps}
        >
          {children}
        </UIText>
      ) : (
        children
      )}
    </Container>
  )
}

export default UIBalloon
