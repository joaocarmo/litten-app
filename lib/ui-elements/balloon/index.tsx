import { Pressable, View } from 'react-native'
import { useTheme } from '@hooks'
import UIText from '@ui-elements/text'
import type { UITextProps } from '@ui-elements/text'
import uiBalloonStyles from '@ui-elements/balloon/index.styles'

export interface UIBalloonProps extends UITextProps {
  type?: 'info' | 'success' | 'warning' | 'error'
}

const UIBalloon = ({
  children,
  style,
  type,
  ...otherProps
}: UIBalloonProps) => {
  const { onPress, onLongPress } = otherProps
  const { createStyles } = useTheme()
  const styles = createStyles(uiBalloonStyles)

  const getColorScheme = () => {
    switch (type) {
      case 'error':
        return {
          container: styles.uiBalloonError,
          text: styles.uiBalloonNormalError,
        }

      case 'info':
        return {
          container: styles.uiBalloonInfo,
          text: styles.uiBalloonNormalInfo,
        }

      default:
        return {
          container: styles.uiBalloonNormal,
          text: styles.uiBalloonNormalText,
        }
    }
  }

  const baloonStyle = getColorScheme()
  const Container =
    typeof onPress === 'function' || typeof onLongPress === 'function'
      ? Pressable
      : View

  return (
    <Container style={[styles.uiBalloon, baloonStyle.container, style]}>
      {typeof children === 'string' ? (
        <UIText
          style={[styles.uiBalloonText, baloonStyle.text]}
          {...otherProps}
        >
          {children}
        </UIText>
      ) : (
        children
      )}
    </Container>
  )
}

UIBalloon.defaultProps = {
  type: '',
}

export default UIBalloon
