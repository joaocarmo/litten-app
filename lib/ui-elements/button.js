/**
 * @format
 * @flow
 */

import type { Node } from 'react'
import { Pressable, StyleSheet, Text } from 'react-native'
import UILoader from 'ui-elements/loader'
import {
  RECOMMENDED_MINIMUM_TAPPABLE_SIZE,
  UI_BUTTON_BORDER_RADIUS,
  UI_BUTTON_FIXED_WIDTH,
  UI_PRESSED_OPACITY,
} from 'utils/constants'
import colors from 'styles/colors'
import { fontSize, fontWeight } from 'styles/typography'

const UIButton: (args: any) => Node = ({
  children,
  compact = false,
  danger = false,
  disabled = false,
  fluid = false,
  loading = false,
  onPress,
  secondary = false,
  style,
  textStyle,
  ...otherProps
}) => {
  const fluidity = fluid ? styles.fluid : styles.notFluid

  const getStyle = () => {
    if (danger) {
      if (disabled) {
        return dangerDisabledStyle
      }
      return dangerStyle
    }
    if (secondary) {
      if (disabled) {
        return secondaryDisabledStyle
      }
      return secondaryStyle
    }
    if (disabled) {
      return primaryDisabledStyle
    }
    return primaryStyle
  }

  return (
    <Pressable
      {...otherProps}
      onPress={disabled ? null : onPress}
      style={({ pressed }) => [
        getStyle(),
        compact ? styles.compact : fluidity,
        style,
        pressed && !disabled && !loading ? styles.pressedStyle : undefined,
      ]}>
      <UILoader active={loading} transparent={100} />
      {!loading && (
        <Text style={StyleSheet.compose(styles.uiButtonText, textStyle)}>
          {children}
        </Text>
      )}
    </Pressable>
  )
}

const styles = StyleSheet.create({
  uiButton: {
    minHeight: RECOMMENDED_MINIMUM_TAPPABLE_SIZE,
    borderRadius: UI_BUTTON_BORDER_RADIUS,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: colors.purple,
  },
  fluid: {
    flex: 1,
    width: '100%',
  },
  notFluid: {
    width: UI_BUTTON_FIXED_WIDTH,
  },
  compact: {
    paddingLeft: 32,
    paddingRight: 32,
  },
  uiButtonText: {
    color: colors.white,
    fontSize: fontSize.large,
    fontWeight: fontWeight.bolder,
  },
  secondary: {
    backgroundColor: colors.blue,
  },
  danger: {
    backgroundColor: colors.red,
  },
  pressedStyle: {
    opacity: UI_PRESSED_OPACITY,
  },
  disabled: {
    opacity: 0.3,
  },
})

const primaryStyle = styles.uiButton
const primaryDisabledStyle = StyleSheet.compose(primaryStyle, styles.disabled)
const secondaryStyle = StyleSheet.compose(primaryStyle, styles.secondary)
const secondaryDisabledStyle = StyleSheet.compose(
  secondaryStyle,
  styles.disabled,
)
const dangerStyle = StyleSheet.compose(primaryStyle, styles.danger)
const dangerDisabledStyle = StyleSheet.compose(dangerStyle, styles.disabled)

export default UIButton
