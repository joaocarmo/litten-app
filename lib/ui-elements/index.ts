import UIAvatar from './avatar'
import UIBadge from './badge'
import UIBalloon from './balloon'
import UIButton from './button'
import UIContainer from './container'
import UIDropdown from './dropdown'
import UIHeader from './header'
import UIIcon from './icon'
import UIImage from './image'
import UIImagePlaceholder from './image-placeholder'
import UIInput from './input'
import UILink from './link'
import UIListItem from './list-item'
import UILoader from './loader'
import UIMessage from './message'
import UIModal from './modal'
import UIOption from './option'
import UIPasswordInput from './password-input'
import UIProgress from './progress'
import UIPrompt from './prompt'
import UISearch from './search'
import UISelect from './select'
import UISeparator from './separator'
import UISetting from './setting'
import UISlider from './slider'
import UISwitch from './switch'
import UIText from './text'
import UITextArea from './textarea'
import UITooltip from './tooltip'

export {
  UIAvatar,
  UIBadge,
  UIBalloon,
  UIButton,
  UIContainer,
  UIDropdown,
  UIHeader,
  UIIcon,
  UIImage,
  UIImagePlaceholder,
  UIInput,
  UILink,
  UIListItem,
  UILoader,
  UIMessage,
  UIModal,
  UIOption,
  UIPasswordInput,
  UIProgress,
  UIPrompt,
  UISearch,
  UISelect,
  UISeparator,
  UISetting,
  UISlider,
  UISwitch,
  UIText,
  UITextArea,
  UITooltip,
}
