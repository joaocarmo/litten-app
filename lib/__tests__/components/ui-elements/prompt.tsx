import TestRenderer from 'react-test-renderer'
import ThemeProvider from 'components/theme/provider'
import { UIPrompt } from 'ui-elements'
beforeAll(() => {
  jest.mock('react-native/Libraries/Utilities/Platform', () => ({
    OS: 'android',
    select: () => null,
  }))
})
describe('Snapshot test for the "UIPrompt" component', () => {
  it('renders correctly', () => {
    const element = TestRenderer.create(
      <ThemeProvider>
        <UIPrompt
          open
          title="Title"
          message="Message"
          type="secure-text"
          cancelLabel="Cancel"
          onCancel={() => undefined}
          confirmLabel="Confirm"
          onConfirm={() => undefined}
          isDestructive
        />
      </ThemeProvider>,
    ).toJSON()
    expect(element).toMatchSnapshot()
  })
})
