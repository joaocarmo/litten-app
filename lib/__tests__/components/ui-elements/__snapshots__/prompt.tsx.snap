// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test for the "UIPrompt" component renders correctly 1`] = `<View />`;
